### PR TITLE
GAWB 2647 - Switcheroo

### PR DESCRIPTION
**Pull request by vkumra-broad**
Changes to do the swticheroo from pet SA to user account.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Make sure Swagger is updated if API changes
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 1]
```diff
@@ -1,15 +1,38 @@
...
   def requireUserInfo: Directive1[UserInfo] = (
     headerValueByName("OIDC_access_token") &
     headerValueByName("OIDC_CLAIM_user_id") &
     headerValueByName("OIDC_CLAIM_expires_in") &
     headerValueByName("OIDC_CLAIM_email")
-  ).tmap { case (token, userId, expiresIn, email) => UserInfo(token, WorkbenchUserId(userId), WorkbenchUserEmail(email), expiresIn.toLong) }
+  ) tflatMap {
+    case (token, userId, expiresIn, email) => {
+      onSuccess(getWorkbenchUserEmailId(email)).map {
+        case Some(resourceType) => UserInfo(token, resourceType.id, resourceType.email, expiresIn.toLong)
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala]: Minor, I don't think `resourceType` is the best name -- maybe `petOwnerUser`?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 2]
```diff
@@ -1,15 +1,38 @@
...
+        case None => UserInfo(token, WorkbenchUserId(userId), WorkbenchUserEmail(email), expiresIn.toLong)
+      }
+    }
+  }
+
+  private def isPetSA(email:String) = email.matches(petSAdomain)
+
+  private def getWorkbenchUserEmailId(email:String):Future[Option[WorkbenchUser]] = {
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala]: Maybe this method should be named `getUserFromPetServiceAccount`

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 3]
```diff
@@ -1,15 +1,38 @@
...
+    if (isPetSA(email))
+      directoryDAO.getUserFromPetServiceAccount(WorkbenchUserServiceAccountEmail(email))
+      else
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala]: Indentation is off

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 4]
```diff
@@ -1,15 +1,38 @@
...
+import org.broadinstitute.dsde.workbench.model._
 import org.broadinstitute.dsde.workbench.sam.model.UserInfo
 
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+
+
 trait StandardUserInfoDirectives extends UserInfoDirectives {
+  val petSAdomain = "\\S+@\\S+.iam.gserviceaccount.com"
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala]: I think you want to escape the dots so they're treated as literal dots, not regex matchers. E.g., right now this matches the following which is not correct:
```
scala> "abc@fooXiamYgserviceaccountZcom".matches(petSAdomain)
res7: Boolean = true
```

This fixes it:
```
scala> val petSAdomain = "\\S+@\\S+\\.iam\\.gserviceaccount\\.com"
petSAdomain: String = \S+@\S+\.iam\.gserviceaccount\.com

scala> "abc@fooXiamYgserviceaccountZcom".matches(petSAdomain)
res8: Boolean = false

scala> "abc@foo.iam.gserviceaccount.com".matches(petSAdomain)
res9: Boolean = true
```

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 5]
```diff
@@ -1,15 +1,38 @@
...
+  ) tflatMap {
+    case (token, userId, expiresIn, email) => {
+      onSuccess(getWorkbenchUserEmailId(email)).map {
+        case Some(resourceType) => UserInfo(token, resourceType.id, resourceType.email, expiresIn.toLong)
+        case None => UserInfo(token, WorkbenchUserId(userId), WorkbenchUserEmail(email), expiresIn.toLong)
+      }
+    }
+  }
+
+  private def isPetSA(email:String) = email.matches(petSAdomain)
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala]: This works, but it causes the pattern to be recompiled each time, which could have a performance impact. You can compile the pattern once like so:

```
val petSAdomain = "\\S+@\\S+\\.iam\\.gserviceaccount\\.com".r

private def isPetSA(email: String) = {
  petSAdomain.pattern.matcher(email).matches
}
```

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/DirectoryDAO.scala [hunk 6]
```diff
@@ -40,4 +40,5 @@ trait DirectoryDAO {
   def getPetServiceAccountForUser(userId: WorkbenchUserId): Future[Option[WorkbenchUserServiceAccountEmail]]
   def addPetServiceAccountToUser(userId: WorkbenchUserId, petServiceAccountEmail: WorkbenchUserServiceAccountEmail): Future[WorkbenchUserServiceAccountEmail]
   def removePetServiceAccountFromUser(userId: WorkbenchUserId): Future[Unit]
+  def getUserFromPetServiceAccount(petSA:WorkbenchUserServiceAccountEmail):Future[Option[WorkbenchUser]]
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/DirectoryDAO.scala]: This is a style thing, but I think we generally prefer to put a space after the colon. So 
```
def getUserFromPetServiceAccount(petSA: WorkbenchUserServiceAccountEmail): Future[Option[WorkbenchUser]]
```

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/MockUserInfoDirectives.scala [hunk 7]
```diff
@@ -1,13 +1,19 @@
...
 import org.broadinstitute.dsde.workbench.sam.model.UserInfo
 
 /**
   * Created by dvoet on 6/7/17.
   */
 trait MockUserInfoDirectives extends UserInfoDirectives {
   val userInfo: UserInfo
-
-  override def requireUserInfo: Directive1[UserInfo] = provide(userInfo)
+  val petSAdomain = "\\S+@\\S+.iam.gserviceaccount.com"
```
Comment (rtitle) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/api/MockUserInfoDirectives.scala]: Same regex comments as above

**Review by rtitle — CHANGES_REQUESTED** [review]
My comments are just nitpicks, overall looks good!

**Review by helgridly — APPROVED** [review]
seems legit but i defer to the 🐶

**Review by rtitle — APPROVED** [review]

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 8]
```diff
@@ -1,15 +1,40 @@
...
+        case None => UserInfo(token, WorkbenchUserId(userId), WorkbenchUserEmail(email), expiresIn.toLong)
+      }
+    }
+  }
+
+  private def getUserFromPetServiceAccount(email:String):Future[Option[WorkbenchUser]] = {
+    if (isPetSA(email))
+      directoryDAO.getUserFromPetServiceAccount(WorkbenchUserServiceAccountEmail(email))
+    else
+      Future(None)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala]: @abaumann it happened again

@vkumra-broad `Future.successful(None)` is a bit better

Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala]: 😕

**Review by dvoet — APPROVED** [review]

**Pull request merged**